app/src/main/java/de/westnordost/streetcomplete/data/osm/edits/upload/ElementEditUploader.kt

```diff
@@ -22,31 +24,50 @@ class ElementEditUploader(
      *  @throws ConflictException if element has been changed server-side in an incompatible way
      */
     suspend fun upload(edit: ElementEdit, getIdProvider: () -> ElementIdProvider): MapDataUpdates {
-        val remoteChanges by lazy { edit.action.createUpdates(RemoteMapDataRepository(mapDataApi), getIdProvider()) }
+        // build changes on top of map data cached in the local database (faster)
         val localChanges by lazy { edit.action.createUpdates(mapDataController, getIdProvider()) }
 
-        val mustUseRemoteData = edit.action::class in ApplicationConstants.EDIT_ACTIONS_NOT_ALLOWED_TO_USE_LOCAL_CHANGES
+        // build changes on top of map data downloaded ad-hoc from remote (additional HTTP requests -> slower)
+        val remoteChanges by lazy { edit.action.createUpdates(RemoteMapDataRepository(mapDataApi), getIdProvider()) }
+
+        // certain edit types don't allow building changes on top of cached map data
+        val mustUseRemoteData = edit.action::class in EDIT_ACTIONS_NOT_ALLOWED_TO_USE_LOCAL_CHANGES
 
-        return if (mustUseRemoteData) {
+        if (!mustUseRemoteData) {
             try {
-                uploadChanges(edit, remoteChanges, false)
-            } catch (e: ConflictException) {
-                // probably changeset closed
-                uploadChanges(edit, remoteChanges, true)
+                return uploadChanges(edit, localChanges, false)
             }
-        } else {
-            try {
-                uploadChanges(edit, localChanges, false)
-            } catch (e: ConflictException) {
-                // either changeset was closed, or element modified, or local element was cleaned from db
+            // either changeset was closed, element modified on remote, element not available in DB
+            catch (e: ConflictException) {
+                // continue execution outside of this if-block (-> try with remote data)
```

Comment:
Maybe this bit can be made more readable by extracting the remote upload handling into a separate function and calling that here and after the `if` block?

Or maybe even extract the local upload handling into a separate function, too?

Reply:
So, I've rewritten it. Take another look! :-)

Reply:
Yeah, that's a little easier to read. Thanks!